### PR TITLE
Vlan: enable vlan ping between guest and host

### PR DESCRIPTION
**Pull request by sitoliu**
id: 1577075
Signed-off-by: Sitong Liu <siliu@redhat.com>

### CODE REVIEW

File: generic/tests/vlan.py [hunk 1]
```diff
@@ -134,6 +137,20 @@ def flood_ping(src, dst):
...
+        Configure vlan over the bridge interface
+        """
+        ip_cmd = "ip link"
+        ifconfig_cmd = "ifconfig"
+        sw_copy_cmd = "%s add link switch name %s type vlan id %s" % (ip_cmd, switch_name, bridge_vlan)
+        try:
+            process.system(sw_copy_cmd)
+            process.system("%s %s %s up" % (ifconfig_cmd, switch_name, switch_ip))
+        except:
+            logging.debug("Failed to configure vlan bridge '%s'" % switch_name)
```
Comment (hereischen) [inline generic/tests/vlan.py]: you don't need this line, you will get this info from `test.error`

File: generic/tests/vlan.py [hunk 2]
```diff
@@ -134,6 +137,20 @@ def flood_ping(src, dst):
                         session=session_flood, timeout=10)
         session_flood.close()
 
+    def bridge_vlan_config():
+        """
+        Configure vlan over the bridge interface
+        """
+        ip_cmd = "ip link"
+        ifconfig_cmd = "ifconfig"
```
Comment (hereischen) [inline generic/tests/vlan.py]:
```python
ip_cmd = "ip link"
ifconfig_cmd = "ifconfig"
```
you can put these two lines directly in your cmds, you don't need to format it into your cmd.

File: generic/tests/vlan.py [hunk 3]
```diff
@@ -218,6 +246,42 @@ def flood_ping(src, dst):
...
+            if vlan == int(bridge_vlan):
+                error_context.context("Run netperf stress test between guests", logging.info)
+                utils_test.run_virt_sub_test(test, params, env, sub_type)
+                if vm.is_dead():
+                    vm.create(params=params)
+
+        error_context.context("Ping test among guests and host", logging.info)
+        guest_vlan = bridge_vlan
+        for vm_index, vm in enumerate(vms):
+            interface = ifname[vm_index] + '.' + str(guest_vlan)
```
Comment (hereischen) [inline generic/tests/vlan.py]: `guest_vlan = bridge_vlan`
`interface = ifname[vm_index] + '.' + str(guest_vlan)`
`bridge_vlan` is a str,  so you don't need to convert it again. 
There is no need to assign `bridge_vlan` to `guest_vlan`. You can use `bridge_vlan` directly, if you want to make it clear that `birdge_vlan` is `guest_vlan`, a comment will be better in this case.

File: generic/tests/vlan.py [hunk 4]
```diff
@@ -218,6 +246,42 @@ def flood_ping(src, dst):
...
+            dest = switch_ip
```
Comment (hereischen) [inline generic/tests/vlan.py]: same with `guest_vlan = bridge_vlan`, a comment maybe more clear.
you can directly use: `err_msg = "%s ping %s unexpected, " % (interface, switch_ip)`

File: generic/tests/vlan.py [hunk 5]
```diff
@@ -218,6 +246,42 @@ def flood_ping(src, dst):
             nc_transfer(test, 0, 1)
             nc_transfer(test, 1, 0)
 
+            if vlan == int(bridge_vlan):
+                error_context.context("Run netperf stress test between guests", logging.info)
+                utils_test.run_virt_sub_test(test, params, env, sub_type)
+                if vm.is_dead():
+                    vm.create(params=params)
```
Comment (hereischen) [inline generic/tests/vlan.py]: Is this an intend error? Which vm is this `vm`?
Can you explain why you create vm when vm is dead?

Comment (sitoliu) [reply generic/tests/vlan.py]: No. Discussed with @xiywang , delete this statements.  It will raise error when sub_test `netperf_stress` fail.

**Comment by sitoliu** [comment]
(1/1) Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.vlan: PASS (1218.08 s)

**Comment by hereischen** [comment]
Acked-by: Haotong Chen hachen@redhat.com

File: generic/tests/vlan.py [hunk 6]
```diff
@@ -153,6 +170,7 @@ def flood_ping(src, dst):
     vms.append(env.get_vm("vm2"))
     for vm_ in vms:
         vm_.verify_alive()
+    bridge_vlan_config()
```
Comment (xiywang) [inline generic/tests/vlan.py]: Could you separate the old cases with your newly added case?
I mean, the old case is refer to RHEL7-95961 and do not need to set switch.
Moreover, in this case, no need to do file transfer test.
So my opinion is to separate them. You can add variants in the cfg.

Comment (xiywang) [reply generic/tests/vlan.py]: And are you going to support Windows guest in this PR?
Or add the support later?

Comment (sitoliu) [reply generic/tests/vlan.py]: Thanks @hereischen @xiywang , will rewrite this case to bridge only vlan test script as Xiyue suggested, please help review it later again. Thanks so much!

**Comment by sitoliu** [comment]
JOB LOG    : /root/avocado/job-results/job-2018-05-29T15.18-4ba6aa9/job.log
 (1/1) Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.bridge_vlan: PASS (637.45 s)

**Comment by xiywang** [comment]
E0602: 69,27: run.set_ip_vlan: Undefined variable 'set_cmd'
Lint: /home/travis/build/autotest/tp-qemu/qemu/tests/bridge_vlan.py FAIL 
/home/travis/build/autotest/tp-qemu/qemu/tests/bridge_vlan.py:14:1: E302 expected 2 blank lines, found 1

would you resubmit the pr?

**Comment by sitoliu** [comment]
Hi @xiywang , I have corrected this issue. The failed reason is
> cmd2 requires Python '>=3.4' but the running Python is 2.7.14

I will resubmit this patch when next update. Thanks.

File: qemu/tests/bridge_vlan.py [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    Test 802.1Q vlan of NIC among guests and host.
+
+    1) Configure vlan over the bridge interface.
+    2) Create two VMs over this bridge.
+    2) Load 8021q module in guest.
```
Comment (pezhang) [inline qemu/tests/bridge_vlan.py]: @sitoliu a simple typo, 2 2)s here.

File: qemu/tests/bridge_vlan.py [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+    mac_str = params.get("mac_str").split(',')
+    os_type = params.get("os_type", "linux")
+    login_timeout = int(params.get("login_timeout", 360))
+
+    linux_modules.load_module("8021q")
+
+    host_vlan_if = "%s.%s" % ("switch", host_vlan_id)
+    if host_vlan_if not in utils_net.get_net_if():
+        host_vlan_if = add_vlan(interface="switch", v_id=host_vlan_id)
+        if host_vlan_if and host_vlan_if in utils_net.get_net_if():
```
Comment (pezhang) [inline qemu/tests/bridge_vlan.py]: @sitoliu  "if host_vlan_if" seems doesn't work here, as it's can not be empty or none.

**Comment by pezhang** [comment]
(1/1) Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.bridge_vlan: PASS (658.56 s)

**Comment by sitoliu** [comment]
@pezhang Modified the code, thanks!

**Comment by pezhang** [comment]
(1/1) Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.bridge_vlan: PASS (657.09 s)

Looks good to me.  Acked-by: Pei Zhang pezhang@redhat.com

File: qemu/tests/bridge_vlan.py [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+        vlan_ip = "%s.%s.%d" % (subnet, host_vlan_id, vm_index + 11)
+        set_ip_vlan(vlan_if, vlan_ip, session=session)
+        vm_vlan_ip.append(vlan_ip)
+
+        error_context.context("Test ping from interface '%s' on guest '%s' to host." %
+                              (vm_vlan_if[vm_index], vm.name), logging.info)
+        ping_vlan(vm, dest=host_vlan_ip, vlan_if=vm_vlan_if[vm_index], session=session)
+        netperf_vlan(client=vm.name, server="localhost")
+
+    error_context.context("Test ping and netperf bwtween guests with interface '%s'" %
```
Comment (xiywang) [inline qemu/tests/bridge_vlan.py]: spelling of between

File: qemu/tests/bridge_vlan.py [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+                          vm_vlan_if[0], logging.info)
+    ping_vlan(vms[0], dest=vm_vlan_ip[1], vlan_if=vm_vlan_if[0], session=sessions[0])
+    netperf_vlan(client=params["main_vm"], server='vm2')
+
+    exithandlers = "exithandlers__%s" % sub_type
+    sub_exit_timeout = int(params.get("sub_exit_timeout", 10))
+    start_time = time.time()
+    end_time = time.time() + float(sub_exit_timeout)
```
Comment (xiywang) [inline qemu/tests/bridge_vlan.py]: end_time = start_time + float(sub_exit_timeout)

Comment (sitoliu) [reply qemu/tests/bridge_vlan.py]: Looks more readable, updated it, thanks.

**Comment by vivianQizhu** [comment]
@xiywang Would you help review it again? Thanks.

File: qemu/tests/bridge_vlan.py [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+    login_timeout = int(params.get("login_timeout", 360))
+
+    linux_modules.load_module("8021q")
+
+    host_vlan_if = "%s.%s" % ("switch", host_vlan_id)
+    if host_vlan_if not in utils_net.get_net_if():
+        host_vlan_if = add_vlan(interface="switch", v_id=host_vlan_id)
+        if host_vlan_if in utils_net.get_net_if():
+            set_ip_vlan(vlan_if=host_vlan_if, vlan_ip=host_vlan_ip)
+            rem_host_vlan_cmd = params["rem_host_vlan_cmd"] % host_vlan_if
```
Comment (xiywang) [inline qemu/tests/bridge_vlan.py]: how about to use rm_xxx instead of rem_xxx?

Comment (vivianQizhu) [reply qemu/tests/bridge_vlan.py]: Yeah, `rm` seems better. And please resolve the CI failure.

File: qemu/tests/bridge_vlan.py [hunk 12]
```diff
@@ -0,0 +1,262 @@
...
+            inter_ip = "%s.%s.%d" % (subnet, host_vlan_id, vm_index + 1)
+            set_ip_vlan(interface, inter_ip, session=session)
+            set_arp_ignore(session)
+            error_context.context("Test ping from guest '%s' to host with "
+                                  "interface '%s'" %
+                                  (vm.name, interface), logging.info)
+            try:
+                ping_vlan(vm, dest=host_vlan_ip, vlan_if=interface,
+                          session=session)
+            except NetPingError as details:
```
Comment (vivianQizhu) [inline qemu/tests/bridge_vlan.py]: `details` is useless, and again, should we fail the test if ping success, with an `else`

Comment (sitoliu) [reply qemu/tests/bridge_vlan.py]: @vivianQizhu Thanks, you are right, will update it.

**Comment by sitoliu** [comment]
Depends on: #1517

File: qemu/tests/cfg/bridge_vlan.cfg [hunk 13]
```diff
@@ -0,0 +1,26 @@
+- bridge_vlan:
+    virt_test_type = qemu libvirt
```
Comment (hereischen) [inline qemu/tests/cfg/bridge_vlan.cfg]: Does this test support tp-libvirt?

**Comment by vivianQizhu** [comment]
@xiywang Please help review, thanks.

**Comment by vivianQizhu** [comment]
@xiywang Would you help provide a test result for it? Thanks.

**Comment by xiywang** [comment]
(1/1) Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.x86_64.myprovider.bridge_vlan: PASS (1050.38 s)

**Comment by xiywang** [comment]
Test result provided in the bugzilla.

**Pull request merged**